src/main/flight/interpolated_setpoint.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "interpolated_setpoint.h"
+#include "build/debug.h"
+#include "fc/rc.h"
+#include "common/maths.h"
+#include "math.h"
+
+#ifdef USE_INTERPOLATED_SP
+static float projectedSetpoint[XYZ_AXIS_COUNT];
+static float prevSetpointSpeed[XYZ_AXIS_COUNT];
+static float prevRawSetpoint[XYZ_AXIS_COUNT];
+static float prevRawDeflection[XYZ_AXIS_COUNT];
+static uint16_t interpolationSteps[XYZ_AXIS_COUNT];
+static float setpointChangePerIteration[XYZ_AXIS_COUNT];
+static float deflectionChangePerIteration[XYZ_AXIS_COUNT];
+static float setpointReservoir[XYZ_AXIS_COUNT];
+static float deflectionReservoir[XYZ_AXIS_COUNT];
+
+// Configuration
+static float ffLookaheadLimit;
+static float ffSpread;
+static float ffMaxRateLimit[XYZ_AXIS_COUNT];
+static float ffMaxRate[XYZ_AXIS_COUNT];
+
+void interpolatedSpInit(const pidProfile_t *pidProfile) {
+    const float ffMaxRateScale = pidProfile->ff_max_rate_limit * 0.01f;
+    ffLookaheadLimit = pidProfile->ff_lookahead_limit * 0.0001f;
+    ffSpread = pidProfile->ff_spread;
+    for (int i = 0; i < XYZ_AXIS_COUNT; i++) {
+        ffMaxRate[i] = applyCurve(i, 1.0f);
+        ffMaxRateLimit[i] = ffMaxRate[i] * ffMaxRateScale;
+    }
+}
+
+FAST_CODE_NOINLINE float interpolatedSpApply(int axis, float pidFrequency, bool newRcFrame) {
+    const float rawSetpoint = getRawSetpoint(axis);
+    const float rawDeflection = getRawDeflection(axis);
+
+    float pidSetpointDelta = 0.0f;
+    static int iterationsSinceLastUpdate[XYZ_AXIS_COUNT];
+    if (newRcFrame) {
+
+        setpointReservoir[axis] -= iterationsSinceLastUpdate[axis] * setpointChangePerIteration[axis];
```

Comment:
I'm sorry, but this has some similarities to a low pass filter, but it's not a low pass filter. The prominent property of the method used in interpolated setpoint ff is that the ff value is kept constant between rc points. No low pass filter does this. It's more comparable to setpoint interpolation on RC which works fairly similar.

Reply:
I mean RC filter running at RC update rate, not PID rate.
Part of last delta is subtracted from new delta - that is what LP filter does.

Current implementation handles case when RC frame is missing (iterationsSinceLastUpdate is larger), but it uses old difference. Using new difference and dividing by (number of missed frames + 1) will probably provide beter results.

Reply:
Not sure. At a high level I look at this feature simply distributing the difference between new setpoint and old setpoint evenly between all loop instances in between. Reservoir has everything that needs to be applied and each application is counted and later subtracted from the reservoir. It seems to me that the old difference needs to be used since that's what we have applied.

Reply:
The reservoir is caring small part of previous RC interval into new one.

If  currentRxRefreshRate / ffSpread is exactly RC update rate, deflectionReservoir would be zero when new RC deflection arrives. Actuall spreading interval is slightly larger, so there will be small part of previous difference (rawDeflection - prevRawDeflection[axis]) remaining in reservoir. This is roughly equivalent to alpha parameter in RC filter (and it is filter with quite high cutoff frequency).

If one RC update is missing, reservoir will be approximately equal to negative of previous difference, so it will correct new difference magnitude (new difference is 2x larger is stick speed is constant). Similar (and more accurate result) may be obtained by dividing deflection difference by time since last available RC update.

This PR is essentially estimating stick speed and setpoint speed + acceleration, but it is doing it in quite complicated way.
It would be best to describe the algorithm in continuous time and then discretize it. Resulting code should be easier to understand and mathematically meaningful ...

Reply:
I missed that `iterationsSinceLastUpdate` is limited to `interpolationSteps` .. When one RC update is missing, setpointReservoir will be exactly zero at this point and new difference will be 2x larger.
That will result in `pidSetpointDelta` `[RC]<based on stick speed>[RC missed] <zero> [RC]<based on double of stick speed>`
This is undesirable. 
It can be mitigated a by setting ffSpread to double of RC update interval, but that will increase RC filtering described above.

Reply:
Yeah sorry - I know how irritating it can be to follow less than well documented code esp if the coder thought about it differently than one would oneself. Would be nice to work out everything from first principles and then write code. Unfortunately that's not what happens in my real life ;-).

Nevertheless you cut to the issues with precision: if packets go missing the results are very poor unless you use ff_spread. And ff_spread lets ff hang around for a bit too long. We do have lpf rc filtering on the output of this to make it a bit smoother.

It's still not worse than what happens today though: if you use normal ff it's based on the 1st derivative of rc setpoint. That becomes a zero also when a point goes missing unless you use punitive amounts of filtering.

Reply:
Yes .. but the code can simply estimate deflection/setpoint speed when RC sample arrives and use this value for some reasonable interval (2-3x RC update frequency).

I was reading this code over and over with each review, now I quite understand it. But the code is complicated and if it get merged, nobody will ever fix it, but probably 'improve' it, making it even more complicated. Eventually it will end up as layer over layer of heuristics, returning to Multiwii days...